### PR TITLE
Implementation of Heart Rate Control Point

### PR DESCRIPTION
**Pull request by g-ortuno**
This PR contains the implementation of the [Heart Rate Control Point](https://developer.bluetooth.org/gatt/characteristics/Pages/CharacteristicViewer.aspx?u=org.bluetooth.characteristic.heart_rate_control_point.xml) characteristic. Since this is a writable characteristic the necessary methods to support writable characteristics are also implemented. After this merge Heart Rate will support `Write`, `Read` and `Notify`.

### CODE REVIEW

File: app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HeartRateServiceFragment.java [hunk 1]
```diff
@@ -232,4 +248,24 @@ private boolean isValidCharacteristicValue(String s, int format) {
...
   }
+
+  @Override
+  public int writeCharacteristic(BluetoothGattCharacteristic characteristic, byte[] value) {
+    // Heart Rate control point is a 8bit characteristic
+    if (value.length != 1) {
+      return BluetoothGatt.GATT_INVALID_ATTRIBUTE_LENGTH;
+    }
+    if ((value[0] & 1) == 1) {
+      mHeartRateMeasurementCharacteristic.setValue(INITIAL_EXPENDED_ENERGY,
```
Comment (jyasskin) [inline app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HeartRateServiceFragment.java]: I think you'll need to do this inside the runOnUiThread also. It's not safe to use an array from two threads concurrently.

File: app/src/main/java/io/github/webbluetoothcg/bletestperipheral/Peripheral.java [hunk 2]
```diff
@@ -130,13 +131,13 @@ public void onCharacteristicReadRequest(BluetoothDevice device, int requestId, i
       super.onCharacteristicReadRequest(device, requestId, offset, characteristic);
       Log.d(TAG, "Device tried to read characteristic: " + characteristic.getUuid());
       Log.d(TAG, "Value: " + Arrays.toString(characteristic.getValue()));
-      if (offset == 0) {
-        mGattServer.sendResponse(device, requestId, BluetoothGatt.GATT_SUCCESS,
-            offset, characteristic.getValue());
-      } else {
+      if (offset != 0) {
```
Comment (jyasskin) [inline app/src/main/java/io/github/webbluetoothcg/bletestperipheral/Peripheral.java]: I would have kept this little refactoring in a separate pull request, but no need to change it now.

File: app/src/main/java/io/github/webbluetoothcg/bletestperipheral/Peripheral.java [hunk 3]
```diff
@@ -144,6 +145,24 @@ public void onNotificationSent(BluetoothDevice device, int status) {
...
+      super.onCharacteristicWriteRequest(device, requestId, characteristic, preparedWrite,
+          responseNeeded, offset, value);
+      Log.v(TAG, "Characteristic Write request: " + Arrays.toString(value));
+      if (responseNeeded) {
+        if (offset != 0) {
+          mGattServer.sendResponse(device, requestId, BluetoothGatt.GATT_INVALID_OFFSET, offset,
+              value);
+          return;
+        }
+        int status = mCurrentServiceFragment.writeCharacteristic(characteristic, value);
```
Comment (jyasskin) [inline app/src/main/java/io/github/webbluetoothcg/bletestperipheral/Peripheral.java]: You should probably write the value regardless of whether a response is needed.

File: app/src/main/java/io/github/webbluetoothcg/bletestperipheral/ServiceFragment.java [hunk 4]
```diff
@@ -6,6 +6,21 @@
...
+  /**
+   * Function to communicate to the ServiceFragment that a device wants to write to a
+   * characteristic.
+   *
+   * The ServiceFragment should check that the value being written is valid and
+   * return a code appropriately. The ServiceFragment should update the UI to reflect the change.
+   * @param characteristic Characteristic to write to
+   * @param value Value to write to the characteristic
+   * @return {@link android.bluetooth.BluetoothGatt#GATT_SUCCESS} if the read operation
+   * was completed successfully.
```
Comment (jyasskin) [inline app/src/main/java/io/github/webbluetoothcg/bletestperipheral/ServiceFragment.java]: Mention what to do if there's an error too.

File: app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HeartRateServiceFragment.java (hunk 1 above)
Comment (jyasskin) [inline app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HeartRateServiceFragment.java]: Does it make sense to split setHeartRateMeasurementValue() into 2 functions so that you can re-use the expended-energy one here? You could set the initial array contents when you create the characteristic.

**Comment by jyasskin** [comment]
LGTM aside from the comments.

File: app/src/main/java/io/github/webbluetoothcg/bletestperipheral/Peripheral.java [hunk 5]
```diff
@@ -144,6 +145,24 @@ public void onNotificationSent(BluetoothDevice device, int status) {
...
+        mGattServer.sendResponse(device, requestId, status, offset, value);
```
Comment (scheib) [inline app/src/main/java/io/github/webbluetoothcg/bletestperipheral/Peripheral.java]: Why send the offset and value back in the sendResponse? Seems sending status alone should be sufficient?

File: app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HeartRateServiceFragment.java (hunk 1 above)
Comment (scheib) [inline app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HeartRateServiceFragment.java]: On Tue, Mar 24, 2015 at 7:49 PM, Jeffrey Yasskin <notifications@github.com>
wrote:

> In
> app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HeartRateServiceFragment.java
> <https://github.com/WebBluetoothCG/ble-test-peripheral-android/pull/29#discussion_r27091634>
> :
>
> > @@ -232,4 +248,24 @@ private boolean isValidCharacteristicValue(String s, int format) {
> >        return false;
> >      }
> >    }
> > +
> > +  @Override
> > +  public int writeCharacteristic(BluetoothGattCharacteristic characteristic, byte[] value) {
> > +    // Heart Rate control point is a 8bit characteristic
> > +    if (value.length != 1) {
> > +      return BluetoothGatt.GATT_INVALID_ATTRIBUTE_LENGTH;
> > +    }
> > +    if ((value[0] & 1) == 1) {
> > +      mHeartRateMeasurementCharacteristic.setValue(INITIAL_EXPENDED_ENERGY,
>
> I think you'll need to do this inside the runOnUiThread also. It's not
> safe to use an array from two threads concurrently.
>

Which array is used on multiple threads?

Comment (g-ortuno) [inline app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HeartRateServiceFragment.java]: Same question as @scheib. Actually I had to add `runOnUiThread` because Android was complaining that the callback was not in the UI thread and it couldn't update `Views` from that thread.

Comment (jyasskin) [inline app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HeartRateServiceFragment.java]: There's an array inside `mHeartRateMeasurementCharacteristic`, but the race would actually be on `mHeartRateMeasurementCharacteristic` itself too. If `writeCharacteristic()` is ever called on a non-UI thread, as the call to `.runOnUiThread()` implies, then the above line would be a write to `mHeartRateMeasurementCharacteristic` on a non-UI thread, while other code in this file modify it in response to things like EditText changes, and while the main activity reads it to respond to BT read requests.

Comment (g-ortuno) [inline app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HeartRateServiceFragment.java]: 1) Gotcha.
2) I agree but shouldn't that kind of change be in a separate PR? This change is about the `Heart Rate Control Point` characteristic.

File: app/src/main/java/io/github/webbluetoothcg/bletestperipheral/Peripheral.java (hunk 2 above)
Comment (g-ortuno) [inline app/src/main/java/io/github/webbluetoothcg/bletestperipheral/Peripheral.java]: Sorry about that.

File: app/src/main/java/io/github/webbluetoothcg/bletestperipheral/Peripheral.java (hunk 3 above)
Comment (g-ortuno) [inline app/src/main/java/io/github/webbluetoothcg/bletestperipheral/Peripheral.java]: Done.

File: app/src/main/java/io/github/webbluetoothcg/bletestperipheral/ServiceFragment.java (hunk 4 above)
Comment (g-ortuno) [inline app/src/main/java/io/github/webbluetoothcg/bletestperipheral/ServiceFragment.java]: Done.

File: app/src/main/java/io/github/webbluetoothcg/bletestperipheral/Peripheral.java (hunk 5 above)
Comment (g-ortuno) [inline app/src/main/java/io/github/webbluetoothcg/bletestperipheral/Peripheral.java]: Done.

**Comment by g-ortuno** [comment]
Ready for review again. @jyasskin @scheib

File: app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HeartRateServiceFragment.java (hunk 1 above)
Comment (jyasskin) [inline app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HeartRateServiceFragment.java]: Re "split `setHeartRateMeasurementValue()` into 2 functions", it can be a separate change, but it'd also make sense in this one since this is the PR that introduces the code duplication.

**Comment by jyasskin** [comment]
LGTM
